can/interfaces/ixxat/canlib_vcinpl2.py

```diff
@@ -0,0 +1,948 @@
+
+import can.util
+
+from . import constants, structures
+from .exceptions import *
+
+__all__ = [
+    "VCITimeout",
+    "VCIError",
+    "VCIBusOffError",
+    "VCIDeviceNotFoundError",
+    "IXXATBus",
+    "vciFormatError",
+]
+
+log = logging.getLogger("can.ixxat")
+
+from time import perf_counter as _timer_function
+
+# Hack to have vciFormatError as a free function, see below
+vciFormatError = None
+
+# main ctypes instance
+_canlib = None
+# TODO: Use ECI driver for linux
+if sys.platform == "win32" or sys.platform == "cygwin":
+    try:
+        _canlib = CLibrary("vcinpl2.dll")
+    except Exception as e:
+        log.warning("Cannot load IXXAT vcinpl library: %s", e)
+else:
+    # Will not work on other systems, but have it importable anyway for
+    # tests/sphinx
+    log.warning("IXXAT VCI library does not work on %s platform", sys.platform)
+
+
+def __vciFormatErrorExtended(library_instance, function, HRESULT, arguments):
+    """Format a VCI error and attach failed function, decoded HRESULT and arguments
+    :param CLibrary library_instance:
+        Mapped instance of IXXAT vcinpl library
+    :param callable function:
+        Failed function
+    :param HRESULT HRESULT:
+        HRESULT returned by vcinpl call
+    :param arguments:
+        Arbitrary arguments tuple
+    :return:
+        Formatted string
+    """
+    # TODO: make sure we don't generate another exception
+    return "{} - arguments were {}".format(
+        __vciFormatError(library_instance, function, HRESULT), arguments
+    )
+
+
+def __vciFormatError(library_instance, function, HRESULT):
+    """Format a VCI error and attach failed function and decoded HRESULT
+    :param CLibrary library_instance:
+        Mapped instance of IXXAT vcinpl library
+    :param callable function:
+        Failed function
+    :param HRESULT HRESULT:
+        HRESULT returned by vcinpl call
+    :return:
+        Formatted string
+    """
+    buf = ctypes.create_string_buffer(constants.VCI_MAX_ERRSTRLEN)
+    ctypes.memset(buf, 0, constants.VCI_MAX_ERRSTRLEN)
+    library_instance.vciFormatError(HRESULT, buf, constants.VCI_MAX_ERRSTRLEN)
+    return "function {} failed ({})".format(
+        function._name, buf.value.decode("utf-8", "replace")
+    )
+
+
+def __check_status(result, function, arguments):
+    """
+    Check the result of a vcinpl function call and raise appropriate exception
+    in case of an error. Used as errcheck function when mapping C functions
+    with ctypes.
+        :param result:
+            Function call numeric result
+        :param callable function:
+            Called function
+        :param arguments:
+            Arbitrary arguments tuple
+        :raise:
+            :class:VCITimeout
+            :class:VCIRxQueueEmptyError
+            :class:StopIteration
+            :class:VCIError
+    """
+    if isinstance(result, int):
+        # Real return value is an unsigned long
+        result = ctypes.c_ulong(result).value
```

Comment:
Removing the line causes a bug, the purpose is to get an unsigned value (sometimes result is negative, and needs to be reinterpreted as unsigned).
Maybe there are other ways to do the same, for example, bitmasking with 0xFFFFFFFF (maybe less portable), but I did not measure the times, so, I think it's safer to leave it as it is.

Reply:
I see, but that just means that the function prototypes are declared incorrectly. See here for example:
```python3
    # HRESULT VCIAPI vciEnumDeviceOpen( OUT PHANDLE hEnum );
    _canlib.map_symbol("vciEnumDeviceOpen", ctypes.c_long, (PHANDLE,), __check_status)
```

The return value must be `HRESULT`, but the prototype is declared with `c_long`.

Reply:
I realized that there is also a ctypes.HRESULT type, but using it was causing errors (probably more adaptions required, I didn't dedicate too much time to investigate).

Using `c_ulong` instead of c_long allows to remove the check.